ci/run.sh

```diff
@@ -100,12 +101,20 @@ test() {
 
 # test dmd
 test_dmd() {
-    # test fewer compiler argument permutations for PRs to reduce CI load
-    if [ "$FULL_BUILD" == "true" ] && [ "$OS_NAME" == "linux"  ]; then
-        make -j1 -C compiler/test MODEL=$MODEL N=$N # all ARGS by default
+    # run the d_do_test unittests
+    $build_path/dmd -m$MODEL -g -i -Icompiler/test/tools -version=NoMain -lowmem -unittest -run compiler/test/tools/d_do_test.d
+
+    # build the run.d helper
+    $build_path/dmd -m$MODEL -g -i -Icompiler/test -release compiler/test/run.d -ofgenerated/run
+
+    # optionally test fewer compiler argument permutations to reduce CI load
+    if [ "$FULL_BUILD" == "true" ] && [ "$OS_NAME" == "linux" ]; then
+        local args=() # use all default ARGS
     else
-        make -j1 -C compiler/test MODEL=$MODEL N=$N ARGS="-O -inline -release"
+        local args=(ARGS="-O -inline -release")
     fi
+
+    generated/run -j$N --environment MODEL=$MODEL HOST_DMD=$build_path/dmd "${args[@]}"
```

Comment:
The Makefile has some fallback logic if it cannot find a D host compiler in PATH, falling back to the freshly generated compiler (`USE_GENERATED` logic). `run.d` doesn't, so currently specifying the fresh compiler here explicitly.

IIRC, Walter has complained about this in the past, as the compiler being tested might not be stable enough to robustly build the test utilities (`d_do_test`, `sanitize_json` etc.), potentially causing misleading failures.

So using the real host compiler here (and probably for building `run.d` too) may make more sense.

Reply:
I tried, but getting various compile errors, tool runtime crashes, unsupported `-i` cmdline flag etc. etc. with various older host compilers - doesn't seem feasible at all for now.

Reply:
I like to use the most recent released compiler as the build tool.